### ISSUE TITLE
Part 2. Regression - Case Study

### ISSUE THREAD

**Issue by XuSShuai** [XuSShuai/Machine-Learning-2017-Fall#2]
## Regression: Output a Scalar

 - stock market forecast
 - self-driver car
 - recommendation

## Example Application

**estimating the combat power(CP) of a pokemon after evolution.**

### Step 1: Model

liner model: $$y = b + \sum{w_ix_i}$$
$y = b + w * x_{cp}$ 
假设根据domain knowledge我们认为进化后的cp值和进化之前的cp($x_{cp}$)值存在线性关系。

### Step 2: Goodness of Function

利用training set，根据Loss function来衡量model中的无数个function的好坏。

$$L(f) = \sum_{i=1}^{N}(\hat{y}^n - f(x_{cp}^{n}))^2$$
$$L(w, b) = \sum_{i=1}^{N}(\hat{y}^n - (b + w*x_{cp}^n))^2$$

### Step 3: Best Function

对最优化问题进行求解：
$$w^{\*}, b^{\*} = argmin_{w, b} \ L(w, b)$$

## Gradient Descent

只要定义了合理的Loss function，并且损失函数对于参数是可微分的，那么就可以使用Gradient Descent来进行optimization problem的求解。

 - 只有一个参数w的时候

![image](https://user-images.githubusercontent.com/29543183/48705995-d2de3580-ec35-11e8-865b-df0c46f438e2.png)

 - 有两个参数w和b的时候

![image](https://user-images.githubusercontent.com/29543183/48706396-1dac7d00-ec37-11e8-8a85-8c33fb3f141a.png)

需要注意的是，不仅仅只有**local minima**和**global minima**处的梯度为0，**saddle point**的梯度也为0，并且梯度的更新在**plateau**上非常慢。

![image](https://user-images.githubusercontent.com/29543183/48706641-d4a8f880-ec37-11e8-8913-fa1e8bb949da.png)

## Model Selection

A more complex model does not always lead to better performance on testing data. This is `overfitting`.

![image](https://user-images.githubusercontent.com/29543183/48708858-b4306c80-ec3e-11e8-91cd-56eb215d2f7f.png)

## Regularization

$$L(w, b) = \sum_{n=1}^{N}\big(\hat{y}^n - (b + \sum w_ix^n_i)\big)^2 + \lambda \sum (w_i)^2$$

现在通过gradient descent系列类的方法最小化loss function的同时，要考虑参数w的影响，所以参数w更小的function会是更好的解，而w很小的function是很平滑的（极端情况下，如果所有的w都是0，那么function是一条水平线），平滑可以理解为当输入有扰动时，输出的变化是很小的（如下两个等式所示，当w小的时候，扰动对结果的影响小）。为什么更加平滑的模型可能是更好的呢？奥卡姆剃刀原理：如无必要，勿增实体。

$$y = b + \sum w_i x_i \longrightarrow y + \sum w_i  \Delta x_i = b + \sum w_i (x_i + \Delta x_i)$$
